tests/libvmi/lifecycle.go

```diff
@@ -1,46 +1,100 @@
 )
 
-func SetupVMI(vmi *v1.VirtualMachineInstance) *v1.VirtualMachineInstance {
+func SetupVMI(vmi *v1.VirtualMachineInstance, loggedInExpecterFactory tests.VMIExpecterFactory) *v1.VirtualMachineInstance {
+	vmiSet := VMISet{"vmi": vmi}
+	vmiSet.Setup(loggedInExpecterFactory)
+	return vmiSet["vmi"]
+}
+
+func CleanupVMI(vmi *v1.VirtualMachineInstance) {
+	vmiSet := VMISet{"vmi": vmi}
+	vmiSet.Cleanup()
+}
+
+type VMISet map[string]*v1.VirtualMachineInstance
+
+func NewVMISet() VMISet {
+	return map[string]*v1.VirtualMachineInstance{}
+}
+
+func (vmis VMISet) Setup(loggedInExpecterFactory tests.VMIExpecterFactory) {
 	virtClient, err := kubecli.GetKubevirtClient()
 	Expect(err).NotTo(HaveOccurred(), "Should initialize KubeVirt client")
 
-	By("Creating the VMI")
-	vmi, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
-	Expect(err).NotTo(HaveOccurred(), "VMI should be successfully created")
-
-	By("Waiting until the VMI gets ready")
-	vmi = tests.WaitUntilVMIReady(vmi, tests.LoggedInCirrosExpecter)
+	for name, vmi := range vmis {
+		By(fmt.Sprintf("Creating VMI %q", name))
+		vmi, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
+		Expect(err).NotTo(HaveOccurred(), "VMI %q should be successfully created", name)
+		vmis[name] = vmi
+	}
 
-	return vmi
+	for name, vmi := range vmis {
+		By(fmt.Sprintf("Waiting until VMI %q gets ready", name))
+		vmis[name] = tests.WaitUntilVMIReady(vmi, loggedInExpecterFactory)
+	}
 }
 
-func CleanupVMI(vmi *v1.VirtualMachineInstance) {
-	if vmi != nil {
-		virtClient, err := kubecli.GetKubevirtClient()
-		Expect(err).NotTo(HaveOccurred(), "Should initialize KubeVirt client")
+func (vmis VMISet) Cleanup() {
+	virtClient, err := kubecli.GetKubevirtClient()
+	Expect(err).NotTo(HaveOccurred(), "Should initialize KubeVirt client")
 
-		By("Deleting the VMI")
-		Expect(virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Delete(vmi.GetName(), &metav1.DeleteOptions{})).To(Succeed())
+	errs := []error{}
+
+	for name, vmi := range vmis {
+		if vmi == nil {
+			continue
+		}
+
+		By(fmt.Sprintf("Deleting VMI %q", name))
+		err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Delete(vmi.GetName(), &metav1.DeleteOptions{})
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	By("Waiting VMIs to be gone")
+	err = wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
+		runningVMIs := []string{}
+
+		for name, vmi := range vmis {
+			if vmi == nil {
+				continue
+			}
 
-		By("Waiting for the VMI to be gone")
-		Eventually(func() error {
 			_, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.GetName(), &metav1.GetOptions{})
-			return err
-		}, 2*time.Minute, time.Second).Should(SatisfyAll(HaveOccurred(), WithTransform(errors.IsNotFound, BeTrue())), "The VMI should be gone within the given timeout")
+			if err == nil {
+				runningVMIs = append(runningVMIs, name)
```

Comment:
if `err == nil` then you can just return `false, nil` right? spare the rest of the loop and the `if` in line 89

Reply:
I'm trying to delete everything we can here, so in the end we report all VMs that we failed to remove while all the rest would be cleaned up. Kind of "best effort" cleanup. Would that explain it?

Reply:
Gotcha